### ISSUE TITLE
Distillation returns NaN under fp16: the 1e-8 normalisation guard is not representable in float16

### ISSUE THREAD

**Issue by 4ktLuffy** [lightonai/pylate#239]
`Distillation` normalises its scores with a `1e-8` epsilon guard:

```python
# pylate/losses/distillation.py
# Avoid division by zero by adding a small epsilon
epsilon = 1e-8
scores = (scores - min_scores) / (max_scores - min_scores + epsilon)
```

**`1e-8` is not representable in float16** — the smallest float16 subnormal is ~5.96e-8, so it rounds to exactly `0.0`:

```python
>>> torch.tensor(1e-8, dtype=torch.float32).item()
1e-08
>>> torch.tensor(1e-8, dtype=torch.bfloat16).item()
1.0011717677116394e-08
>>> torch.tensor(1e-8, dtype=torch.float16).item()
0.0
```

Under `autocast(float16)` the guard disappears, so a candidate group whose scores are equal divides by zero.

### Reproduction

`losses.Distillation(model=...)` with defaults (`normalize_scores=True`), one query and four documents a teacher scores identically:

| autocast | loss | parameters with NaN gradients |
|---|---|---|
| none (fp32) | 0.006223 | — |
| bfloat16 | 0.002742 | — |
| **float16** | **NaN** | **135 / 135** |

`normalize_scores=False` avoids it, which confirms the normalisation is the source. `examples/train/knowledge_distillation.py` sets `fp16=True`.

The same code is at `pylate/evaluation/colbert_distillation.py:180`, where it produces a NaN evaluation metric by the same mechanism.

### The obvious fix does not work

This is the part worth a reviewer's attention. Normalising the forward pass in fp32 makes the **loss** correct while leaving every gradient NaN:

| variant | fp16 loss | NaN gradients |
|---|---|---|
| `epsilon = 1e-8` (current) | NaN | 135/135 |
| cast the forward to fp32 | 0.006223 — looks correct | **135/135 — still broken** |
| `(max - min).clamp(min=torch.finfo(dtype).tiny)` | 0.006648 | **0/135** |

With a near-zero denominator the gradient scales as `1/epsilon ≈ 1e8`, which exceeds float16's maximum of 65504 and becomes `inf`; `inf * 0` is then `NaN` in the backward pass. Clamping to `finfo.tiny` (6.1e-5 in float16) keeps the gradient near 1.6e4, inside range, and fixes both the loss and the gradients.

### How often this triggers

Measured on real scores rather than assumed — SciFact, 1,500 documents, 300 queries, `lightonai/GTE-ModernColBERT-v1` (score range 7.6–46.9), grouping each query's top-k as a candidate set:

| group size k | groups collapsing to a single float16 value |
|---|---|
| 2 | 10 / 300 (3.3%) |
| 4, 8, 16, 32 | 0 / 300 |

So it fires on small candidate groups rather than wide ones. Worth noting the `Distillation` docstring example uses exactly two documents. 44/300 queries have their top-2 within one float16 ulp (0.031 at score ~32).

### Severity, bounded honestly

`GradScaler` detects the NaN gradients and skips the optimizer step, so under `Trainer(fp16=True)` **the weights are not corrupted** — the cost is a silently skipped step and a NaN in the loss logs. Code calling the loss under a manual `torch.autocast` without a scaler does get NaN weights.

### Not verified

- No CUDA hardware here. The mechanism is dtype arithmetic and so device-independent, but the numbers above are from CPU and MPS.
- Whether the KD datasets used in training (`lightonai/ms-marco-en-bge`) contain collapsing groups — only that SciFact-derived groups do at k=2.
- The clamp removes the additive `1e-8` in fp32 as well; I have not measured whether that shifts any published number.

Happy to open the PR — fix at both sites plus a regression test that fails on float16 only.
